vala: Port Vala to Meson

Diego-Ivan opened the pull request:
This PR ports Vala from scripts to Meson, and introduces a minimal Vala template with the Workbench API file and a meson build file.

Fixes https://github.com/workbenchdev/Workbench/issues/964
Fixes #913

Diego-Ivan committed:
vala: Create Vala template

Initial work to port Vala to Meson

Diego-Ivan committed:
vala: Compile and run using Meson

Diego-Ivan committed:
vala: Setup Vala project before starting the language server

Closes #913

Diego-Ivan committed:
cli: Use Meson in Vala checks

sonnyp reviewed (commented):
Thanks for working on this!

One thing to keep in mind is we need to keep compiling existing Vala project so please make sure that's working

I will let @lw64 do the first review

Diego-Ivan commented:
@sonnyp Existing sessions and all the demos work as expected. Projects don't, because meson won't accept the file paths given by a portal. I tested this with Rust/cargo and it happens too.

A possible solution might be copying the Vala sources to a temporary folder that meson accepts (like the cache directory) and compile it there. The source files are usually small so it shouldn't be expensive. What do you think?

Diego-Ivan commented:
Same as #846 unfortunately :(

Diego-Ivan commented:
> I'm also interested to avoid having a builddir in the project directory.

I think it would be better too, since Meson's builddirs can break between major versions, and old projects may break as a consequence of this.

I'll give it a try this week :)

Diego-Ivan committed:
util: Add createTmpDir function

This function will allow to create temporary directory. This will
allow, for instance, meson to create a builddir that will not be
stored inside a session.

Diego-Ivan committed:
vala: Compile inside a temporary directory

This allows to compile projects without granting access to the filesystem.
Additionally, all sessions opened that compile Vala code will use
the same build directory, allowing us to reuse the same meson
configuration.

Diego-Ivan commented:
I think this PR is ready! A little summary of the changes:

* Vala projects will be built in the `/tmp/vala` directory, instead of the working directory of the session.
* The `/tmp/vala` folder will contain all the files necessary to compile a session. Every time _Run_ is clicked, the `main.vala` file in the tmp folder will be replaced by the demo's/project's `main.vala`. Probably we can do this with Rust too, so we don't have to grant filesystem permissions to compile projects.

Also updated the newer demos in https://github.com/workbenchdev/demos/pull/187, tested locally and CI passed, so this is ready for review.

sonnyp commented on src/cli/main.js:
```diff
@@ -352,16 +352,30 @@ async function ci({ filenames }) {
-        "workbench.vala",
-      );
-      file_api.copy(
+      const current_dir = Gio.File.new_for_path(GLib.get_current_dir());
+
+      const template_dir = GLib.getenv("FLATPAK_ID")
+        ? Gio.File.new_for_path(
+            `/app/share/${GLib.getenv("FLATPAK_ID")}/langs/vala/template`,
+          )
+        : current_dir.resolve_relative_path("src/langs/vala/template");
+
+      const api_file = template_dir.get_child("workbench.vala");
+      api_file.copy(
```
```suggestion
      const template_dir = Gio.File.new_for_path(pkg.pkgdatadir).get_child("langs/vala/template");

      template_dir.get_child("workbench.vala").copy(
```

sonnyp commented on src/langs/vala/Compiler.js:
```diff
@@ -60,11 +90,3 @@ export default function ValaCompiler({ session }) {
 
   return { compile, run };
 }
-
-// Takes a string starting with the line
-// #!/usr/bin/env -S vala workbench.vala --pkg gtk4 --pkg libadwaita-1
-// and return ["--pkg", "gtk4", "--pkg", "libadwaita-1"]
-// FIXME: consider using https://docs.gtk.org/glib/struct.OptionContext.html instead
-function getValaCompilerArguments(text) {
-  return text.split("\n")[0]?.split("-S vala ")[1]?.split(" ") || [];
-}
```
:tada:

sonnyp commented on src/langs/vala/Compiler.js:
```diff
@@ -1,49 +1,79 @@
-      "--vapi",
-      "/dev/null",
-      ...args,
+    const meson_clean = meson_launcher.spawnv([
```
why cleaning?

Diego-Ivan replied:
When replacing the file in the build dir, meson will not detect any changes on the file, thus not compiling anything and just running the project right away. This is a problem when a user has two open projects and working on them simultaneously, so I went for cleaning the build files and recompiling. I thought this approach was OK since we don't really compile much.

sonnyp commented on src/langs/vala/Compiler.js:
```diff
@@ -1,49 +1,79 @@
+import { createTmpDir, copy } from "../../util.js";
+import { setupValaProject } from "./vala.js";
+
+let ready_to_build = false;
```
I'm confused about why this is needed

Diego-Ivan replied:
I did this to know if the Vala builddir was already setup (as it needs a meson file and the workbench API). This could be done checking if a build file (like `workbench.vala`) exists, but I went for this instead as it is one less I/O operation.

sonnyp commented on src/langs/vala/template/meson.build:
```diff
@@ -0,0 +1,39 @@
+
+dependencies = [
+    m_dep,
+    dependency('libadwaita-1'),
+    dependency('shumate-1.0'),
+    dependency('libportal-gtk4'),
+    dependency('libsoup-3.0'),
+    dependency('webkitgtk-6.0'),
+    dependency('gstreamer-1.0'),
+    dependency('gtksourceview-5'),
+    dependency('json-glib-1.0'),
```
```suggestion
    dependency('libadwaita-1'),
    dependency('shumate-1.0'),
    dependency('libportal-gtk4'),
    dependency('libsoup-3.0'),
    dependency('webkitgtk-6.0'),
    dependency('gstreamer-1.0'),
    dependency('gtksourceview-5'),
    dependency('json-glib-1.0'),
    dependency('libspelling-1'),
```

sonnyp commented on src/langs/vala/vala.js:
```diff
@@ -2,22 +2,16 @@ import Gio from "gi://Gio";
-    null,
-    null,
-  );
+  // VLS needs the project to be already setup once it starts,
+  // otherwise it won't pick it up later.
+  setupValaProject(file.get_parent()).catch(console.error);
```
is it possible to move the meson and workbench.vala files under a subdir? to avoid cluttering the project

and we might need `meson.build` for other purposes too

Diego-Ivan replied:
Sure, I think this is possible. Would that subdir still be inside the project folder? Or do you mean saving *just* the `main.vala` in the project directory and get rid of all of the build stuff? I'm guessing this last option is also possible.

sonnyp commented on src/util.js:
```diff
@@ -211,3 +211,25 @@ export async function copy(filename, source_dir, dest_dir, flags) {
+
+const tmp_dirs = {};
+
+export async function createTmpDir(filename) {
```
checkout 

* https://gjs-docs.gnome.org/gio20~2.0/gio.file#function-new_tmp
* https://docs.gtk.org/glib/type_func.Dir.make_tmp.html

Diego-Ivan replied:
Thanks, I'll check them out!
